File: lib/src/image_captioning_assistant/generate/bias_analysis/find_biases_in_short_work.py
# Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
# Terms and the SOW between the parties dated 2025.

"""Generate bias analysis for an image."""

import json
import logging
from typing import Any

import boto3
from pydantic_core import ValidationError

import image_captioning_assistant.generate.prompts as p
from image_captioning_assistant.data.data_classes import WorkBiasAnalysis
from image_captioning_assistant.generate.bias_analysis.utils import create_messages, load_and_resize_images
from image_captioning_assistant.generate.utils import (
    extract_json_and_cot_from_text,
    format_request_body,
    LLMResponseParsingError,
)

logger = logging.getLogger(__name__)


def find_biases_in_short_work(
    image_s3_uris: list[str],
    s3_kwargs: dict[str, Any],
    llm_kwargs: dict[str, Any],
    resize_kwargs: dict[str, Any],
    work_context: str | None = None,
    original_metadata: str | None = None,
    bedrock_runtime: Any | None = None,
) -> WorkBiasAnalysis:
    """Find biases in one or two images and, optionally, their existing metadata."""
    #     # connect to runtime if not provided
    if bedrock_runtime is None:
        if "region_name" in llm_kwargs:
            bedrock_runtime = boto3.client("bedrock-runtime", region_name=llm_kwargs["region_name"])
        else:
            bedrock_runtime = boto3.client("bedrock-runtime")
    # Load and resize image bytes
    if len(image_s3_uris) > 2:
        return RuntimeError("maximum of 2 images (front and back) supported for short work")
    if len(image_s3_uris) > 0:
        img_bytes_list = load_and_resize_images(image_s3_uris, s3_kwargs, resize_kwargs)
    else:
        img_bytes_list = []
    # Create messages and request body
    model_name = llm_kwargs["model_id"]
    messages = create_messages(
        img_bytes_list=img_bytes_list,
        work_context=work_context,
        original_metadata=original_metadata,
        model_name=model_name,
    )
    court_order = False
    # Retry loop for robustness around structured metadata
    for attempt in range(5):
        try:
            request_body = format_request_body(model_name, messages, court_order=court_order)
            response = bedrock_runtime.invoke_model(modelId=model_name, body=json.dumps(request_body))

            # Process the response
            result = json.loads(response["body"].read())
            if "claude" in model_name:
                llm_output = result["content"][0]["text"]
            elif "nova" in model_name:
                llm_output = result["output"]["message"]["content"][0]["text"]
            else:
                raise ValueError("ModelId " + model_name + " not supported, must be set up")

            # parse output and log chain of thought
            cot, json_dict = extract_json_and_cot_from_text(llm_output)
            logger.info(f"\n\n********** CHAIN OF THOUGHT **********\n {cot} \n\n")
            # validate correct number of biases output
            if len(image_s3_uris) != len(json_dict["page_biases"]):
                raise LLMResponseParsingError(f"incorrect number of bias lists for {len(image_s3_uris)} pages")
            return WorkBiasAnalysis(**json_dict)

        except Exception as e:
            logger.warning(f"Attempt {attempt+1}/5 failed: {str(e)}")
            if attempt == 4:
                # need to raise exception that was thrown for debugging purposes as invocation is in try block
                raise e

            if isinstance(e, LLMResponseParsingError) or isinstance(e, ValidationError):
                raw_output = llm_output.split(p.COT_TAG_END)[-1]
                logger.debug("Raw model output:", raw_output)
                if (
                    "apologize" in raw_output.lower()
                    or "i cannot" in raw_output.lower()
                    or "i can't" in raw_output.lower()
                ):
                    court_order = True

    raise RuntimeError("Failed to parse model output after 5 attempts")

File: lib/ruby/src/image_captioning_assistant/generate/bias_analysis/find_biases_in_short_work.rb
# Copyright © Amazon.com and Affiliates: This deliverable is considered Developed Content as defined in the AWS Service
# Terms and the SOW between the parties dated 2025.

require 'aws-sdk-bedrockruntime'
require 'json'

module ImageCaptioningAssistant
  module Generate
    module BiasAnalysis
      def self.find_biases_in_short_work(
        image_s3_uris:,
        s3_kwargs:,
        resize_kwargs:,
        llm_kwargs:,
        work_context: nil,
        original_metadata: nil,
        bedrock_runtime: nil
      )
        if bedrock_runtime.nil?
          bedrock_runtime = if llm_kwargs[:region]
                              Aws::BedrockRuntime::Client.new(region: llm_kwargs[:region])
                            else
                              Aws::BedrockRuntime::Client.new
                            end
        end

        raise 'maximum of 2 images (front and back) supported for short work' if image_s3_uris.length > 2

        img_bytes_list = []
        if image_s3_uris.length > 0
          img_bytes_list = BiasAnalysis.load_and_resize_images(
            image_s3_uris: image_s3_uris,
            s3_kwargs: s3_kwargs,
            resize_kwargs: resize_kwargs
          )
        end

        model_name = llm_kwargs[:model_id]
        messages = create_messages(
          img_bytes_list: img_bytes_list,
          work_context: work_context,
          original_metadata: original_metadata,
          model_name: model_name
        )

        court_order = false
        llm_output = nil

        5.times do |attempt|
          request_body = Utils.format_request_body(model_name, messages, court_order: court_order)

          response = bedrock_runtime.invoke_model({
                                                    model_id: model_name,
                                                    body: JSON.generate(request_body)
                                                  })

          result = JSON.parse(response.body.read)
          llm_output = if model_name.include?('claude')
                         result['content'][0]['text']
                       elsif model_name.include?('nova')
                         result['output']['message']['content'][0]['text']
                       else
                         raise ArgumentError, "ModelId #{model_name} not supported"
                       end

          cot, json_dict = Utils.extract_json_and_cot_from_text(llm_output)
          Utils::LOGGER.debug("\n\n********** CHAIN OF THOUGHT **********\n #{cot} \n\n")

          if !image_s3_uris.empty? && image_s3_uris.length != json_dict['page_biases'].length
            raise Utils::LLMResponseParsingError, "incorrect number of bias lists for #{image_s3_uris.length} pages"
          end

          if json_dict['metadata_biases'] && json_dict['metadata_biases']['biases']
            json_dict['metadata_biases']['biases'].each do |bias|
              bias['level'] = bias['level'].to_s.downcase if bias['level']
              bias['type'] = bias['type'].to_s.downcase if bias['type']
            end
          end

          json_dict['page_biases'].each do |page|
            next unless page && page['biases']

            page['biases'].each do |bias|
              bias['level'] = bias['level'].to_s.downcase if bias['level']
              bias['type'] = bias['type'].to_s.downcase if bias['type']
            end
          end

          return Data::WorkBiasAnalysis.new(
            metadata_biases: json_dict['metadata_biases'],
            page_biases: json_dict['page_biases']
          )
        rescue StandardError => e
          Utils::LOGGER.warn("Attempt #{attempt + 1}/5 failed: #{e.message}")
          raise e if attempt == 4

          if e.is_a?(Utils::LLMResponseParsingError) || e.is_a?(ArgumentError)
            raw_output = llm_output.to_s.split(Prompts::COT_TAG_END)[-1].to_s.downcase
            court_order = true if raw_output =~ /apologize|i cannot|i can't/
          end
        end

        raise 'Failed to parse model output after 5 attempts'
      end

      def self.create_messages(img_bytes_list:, model_name:, work_context: nil, original_metadata: nil)
        prompt = Prompts::USER_PROMPT_BIAS

        if work_context || original_metadata
          context_str = "Additional Context:\n"
          context_str += "Work Context: #{work_context}\n" if work_context
          context_str += "Original Metadata: #{original_metadata}\n" if original_metadata
          prompt = "#{prompt}\n\n#{context_str}"
        end

        if model_name.include?('claude')
          Utils.format_prompt_for_claude(
            prompt: prompt,
            img_bytes_list: img_bytes_list,
            assistant_start: Prompts::COT_TAG
          )
        elsif model_name.include?('nova')
          Utils.format_prompt_for_nova(
            prompt: prompt,
            img_bytes_list: img_bytes_list,
            assistant_start: Prompts::COT_TAG
          )
        else
          raise ArgumentError, "model #{model_name} not supported"
        end
      end
    end
  end
end
